flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2083,12 +2205,50 @@ class ArrayExprLowering {
                   }
                 } else {
                   // A regular scalar index, which does not yield an array
-                  // section. Use a degenerate slice operation `(e:e:1)` in this
-                  // dimension as a placeholder.
+                  // section. Use a degenerate slice operation `(e:undef:undef)`
+                  // in this dimension as a placeholder. This does not
+                  // necessarily change the rank of the original array, so the
+                  // iteration space must also be extended to include this
+                  // expression in this dimension to adjust to the array's
+                  // declared rank.
+                  auto base = x.base();
+                  ScalarExprLowering sel{loc, converter, symMap, stmtCtx};
+                  auto exv = base.IsSymbol() ? sel.gen(base.GetFirstSymbol())
+                                             : sel.gen(base.GetComponent());
                   auto v = fir::getBase(asScalar(e));
                   trips.push_back(v);
-                  trips.push_back(v);
-                  trips.push_back(one);
+                  auto undef = builder.create<fir::UndefOp>(loc, idxTy);
```

Comment:
This is a special placeholder encoding. (It replaces the `e:e:1` encoding used earlier.) Since `undefined` is FIR and not Fortran, there should never be the case that a Fortran expression's semantics are "undefined".
Your idea of using an unsupported type is a good one though.